cachito/workers/pkg_managers/pip.py

```diff
@@ -1203,3 +1208,264 @@ def finalize_nexus_for_pip_request(pip_repo_name, raw_repo_name, username):
+            ignored.append(option)
+
+    if ignored:
+        msg = f"Cachito will ignore the following options: {', '.join(ignored)}"
+        log.info(msg)
+
+    if rejected:
+        msg = f"Cachito does not support the following options: {', '.join(rejected)}"
+        raise ValidationError(msg)
+
+    return {
+        "require_hashes": require_hashes,
+    }
+
+
+def _validate_requirements(requirements, require_hashes):
+    """
+    Validate that all requirements meet Cachito expectations.
+
+    :param list[PipRequirement] requirements: All requirements from a file
+    :param bool require_hashes: True if all requirements must specify a checksum
+    :raise ValidationError: If any requirement does not meet expectations
+    """
+    # Fail if any PyPI dependency is not pinned to an exact version
+    for req in filter(lambda r: r.kind == "pypi", requirements):
+        vspec = req.version_specs
+        if len(vspec) != 1 or vspec[0][0] not in ("==", "==="):
+            msg = f"Requirement must be pinned to an exact version: {req.download_line}"
+            raise ValidationError(msg)
+
+    # Fail if any dependency requires a hash but does not specify one
+    for req in requirements:
+        hash_required = require_hashes or req.kind == "url"
+
+        if hash_required and not req.hashes and not req.qualifiers.get("cachito_hash"):
+            msg = f"Hash is required, dependency does not specify any: {req.download_line}"
+            raise ValidationError(msg)
+
+
+def _download_pypi_package(requirement, pip_deps_dir, pypi_url, pypi_auth):
+    """
+    Download the sdist (source distribution) of a PyPI package.
+
+    The package must be pinned to an exact version using the '==' (or '===') operator.
+    While the specification defines the '==' operator as slightly magical (reference:
+    https://www.python.org/dev/peps/pep-0440/#version-matching), we treat the version
+    as exact.
+
+    Does not download any dependencies (implied: ignores extras). Ignores environment
+    markers (target environment is not known to Cachito).
+
+    :param PipRequirement requirement: PyPI requirement from a requirement.txt file
+    :param Path pip_deps_dir: The deps/pip directory in a Cachito request bundle
+    :param str pypi_url: URL of PyPI (proxy) server
+    :param requests.auth.AuthBase pypi_auth: Authorization for the PyPI server
+
+    :return: Path to downloaded file
+    """
+    package = requirement.package
+    version = requirement.version_specs[0][1]
+
+    # See https://warehouse.readthedocs.io/api-reference/json/
+    package_url = f"{pypi_url.rstrip('/')}/pypi/{package}/{version}/json"
+    try:
+        pypi_resp = requests_session.get(package_url, auth=pypi_auth)
+        pypi_resp.raise_for_status()
+    except requests.RequestException as e:
+        raise CachitoError(f"PyPI query failed: {e}")
+
+    data = pypi_resp.json()
+    sdists = [pkg for pkg in data["urls"] if pkg["packagetype"] == "sdist"]
```

Comment:
Should we be using ` sdists = [pkg for pkg in data.get("urls", []) if pkg.get("packagetype", None) == "sdist"]` just to be safer as this is a response coming down from a request?

Reply:
We are already calling `raise_for_status()`  for the response and the call to `json()` should raise a ValueError` in case there is no JSON object in the response.

+1 to being safe here (we are also relying on the Nexus implementation here)

Reply:
Done, also made a similar change for `sdist["yanked"]`. Will leave `sdist["url"]` and `sdist["filename"]` as is, we have no choice but to rely on them